### PR TITLE
Added chunking/compression for string-only objects

### LINKED ISSUE #1041: [Feature]: Add Compression support for compound objects
### What would you like to see added to NeuroConv?

`DatasetIOConfiguration.from_neurodata_object` currently raises a `NotImplementedError` for compound objects with dtype 'object'. 

This makes it very difficult to use the `KiloSortSortingInterface`, which includes a compound object for unit quality ('good' or 'noise').

### Is your feature request related to a problem?

_No response_

### Do you have any interest in helping implement the feature?

Yes.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/catalystneuro/neuroconv/blob/main/.github/CODE_OF_CONDUCT.rst)
- [X] Have you ensured this bug was not already [reported](https://github.com/catalystneuro/neuroconv/issues)?

### PR DESCRIPTION
**Pull request by pauladkisson**
Addresses #1041

### CODE REVIEW

**Comment by h-mayorquin** [comment]
OK, this should not be throwing an error. Can you just pass a configuration with no chukning or compression for those objects?

I also need to take a look on the gin dataset from phy.

**Comment by pauladkisson** [comment]
> Can you just pass a configuration with no chukning or compression for those objects?

Passed in no compression (chunk_shape is required to pass validation, so I kept that) and removed the warning.

**Comment by h-mayorquin** [comment]
Yeah, I remember that I discussed that issue about the chunk_shape and full_shape being mandatory. I think that the argument was that we should not being creating configurations for those cases at all. Does that ring a bell, @CodyCBakerPhD ?

I think his idea was it should be skipped at a higher level like `get_default_dataset_io_configurations` where  `from_neurodata_object` is called.

**Comment by pauladkisson** [comment]
> I think that the argument was that we should not being creating configurations for those cases at all.

I really don't see why we should disallow this, esp since the docs seem to imply that setting these properties to None is the way to disable chunking/compression while still providing a backendConfiguration: https://neuroconv.readthedocs.io/en/main/user_guide/backend_configuration.html

**Comment by h-mayorquin** [comment]
OK, I am weary of this when if the array is too large because we will be creating chunks that are too large but I think that this will be an edge case. In a future PR let's think about how to differentiate objects that are strings (the motivation of this PR) from things that are real objects (I am not even sure we can write the latter to NWB)

Can you add a test for this for this behavior?

**Comment by pauladkisson** [comment]
> OK, I am *wary of this when if the array is too large because we will be creating chunks that are too large but I think that this will be an edge case.

Maybe we should keep the warning then?

**Comment by pauladkisson** [comment]
> Can you add a test for this for this behavior?

Done!

**Comment by h-mayorquin** [comment]
This looks good. 

Here you are testing for arrays that have object type but are in fact string arrays. I would like to hear your thoughts about this comment: https://github.com/catalystneuro/neuroconv/issues/1041#issuecomment-2332668040. I want to move the code in that direction. Would you like to help? It can be here or in another PR but your test will need to be modified a bit. Otherwise, I can take care of it in another PR.


Also, can you unlink the issue so it is not closed when we merge?

File: src/neuroconv/tools/nwb_helpers/_configuration_models/_base_dataset_io.py [hunk 1]
```diff
@@ -264,6 +265,21 @@ def from_neurodata_object(cls, neurodata_object: Container, dataset_name: Litera
         full_shape = get_data_shape(data=candidate_dataset)
         dtype = _infer_dtype(dataset=candidate_dataset)
 
+        if dtype == np.dtype(
```
Comment (h-mayorquin) [inline src/neuroconv/tools/nwb_helpers/_configuration_models/_base_dataset_io.py:268]: I would add this logic inside the last if in the branch below so it is a sub-case of object

Comment (h-mayorquin) [reply src/neuroconv/tools/nwb_helpers/_configuration_models/_base_dataset_io.py:268]: I mean here:
https://github.com/catalystneuro/neuroconv/blob/562bf7ff5222ea6d90f8ba6b4f440b3afcd64078/src/neuroconv/tools/nwb_helpers/_configuration_models/_base_dataset_io.py#L295-L302

**Review by h-mayorquin — APPROVED** [review]
LGTM

**Comment by h-mayorquin** [comment]
Cody brought a new case to me that I want to consider before merging this. I will do some experimentation.

File: src/neuroconv/tools/nwb_helpers/_configuration_models/_base_dataset_io.py [hunk 2]
```diff
@@ -277,10 +278,33 @@ def from_neurodata_object(cls, neurodata_object: Container, dataset_name: Litera
...
+                )
+                compression_method = "gzip"
+            else:
+                chunk_shape = full_shape  # validate_all_shapes fails if chunk_shape or buffer_shape is None
+                buffer_shape = full_shape
+                compression_method = None
+                warnings.warn(
+                    f"Default chunking and compression options for compound objects are not optimized. "
+                    f"Consider manually specifying DatasetIOConfiguration for dataset at '{location_in_file}'."
+                )
```
[the remark this answers is not in the document]
Comment (h-mayorquin) [reply src/neuroconv/tools/nwb_helpers/_configuration_models/_base_dataset_io.py:307]: It is the same issue. If you use the latest version of spikeinterface (to be released soon) that problem should be solved.

File: src/neuroconv/tools/nwb_helpers/_configuration_models/_base_dataset_io.py [hunk 3]
```diff
@@ -277,10 +278,33 @@ def from_neurodata_object(cls, neurodata_object: Container, dataset_name: Litera
...
             compression_method = "gzip"
         elif dtype == np.dtype("object"):  # Unclear what default chunking/compression should be for compound objects
-            raise NotImplementedError(
-                f"Unable to create a `DatasetIOConfiguration` for the dataset at '{location_in_file}'"
-                f"for neurodata object '{neurodata_object}' of type '{type(neurodata_object)}'!"
-            )
+            # pandas reads in strings as objects by default: https://pandas.pydata.org/docs/user_guide/text.html
+            max_string_length = 1
+            all_elements_are_strings = True
+            for element in candidate_dataset[:].flat:
```
Comment (h-mayorquin) [inline src/neuroconv/tools/nwb_helpers/_configuration_models/_base_dataset_io.py:284]: You can do this kind of thing with list comprehension which are optimized and I also think is more readable.

File: src/neuroconv/tools/nwb_helpers/_configuration_models/_base_dataset_io.py [hunk 4]
```diff
@@ -277,10 +278,33 @@ def from_neurodata_object(cls, neurodata_object: Container, dataset_name: Litera
...
+                if not isinstance(element, str):
+                    all_elements_are_strings = False
+                    break
+                string_length = len(element)
+                if string_length > max_string_length:
```
Comment (h-mayorquin) [inline src/neuroconv/tools/nwb_helpers/_configuration_models/_base_dataset_io.py:289]: Instead of estimating yourself the dtype,  I would get do numpy do the casting `np.array(data).dtype` and get the type from them. It would be less brittle I feel.

**Review by h-mayorquin — CHANGES_REQUESTED** [review]
OK, I changed my mind about the compound objects for three reasons:
1)  I don't have time to test the troubling example this week and I need this https://github.com/catalystneuro/neuroconv/pull/1067 first.
2) I would do another patch release before the minor release. 

For the patch release I would like to include the "compound objects that are actually strings" changes of this PR.

I would suggest do the following:

Let's keep the code for objects that are real strings but still keep the previous behavior (throw an error) for compound types. 

This could be achieved by changing the warning to an error and temporarily commenting out the corresponding tests

I can take care of these modifications @pauladkisson but if you want to do it that's fine as well.

**Comment by pauladkisson** [comment]
> For the patch release I would like to include the "compound objects that are actually strings" changes of this PR.
> I would suggest do the following:
> Let's keep the code for objects that are real strings but still keep the previous behavior (throw an error) for compound types.
> This could be achieved by changing the warning to an error and temporarily commenting out the corresponding tests

tbh I'd rather just wait and take care of all compound objects (including the problematic mixed str-NaN case) in this PR.
Is there a reason you need it in the upcoming patch?

**Comment by pauladkisson** [comment]
> In short: I want to have the bug solved and I want to think about the complicated case more. They are two different issues so I would prefer them to address them piece-wise.

Ok, I am conviced. 👍

File: src/neuroconv/tools/nwb_helpers/_configuration_models/_base_dataset_io.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (h-mayorquin) [reply src/neuroconv/tools/nwb_helpers/_configuration_models/_base_dataset_io.py:307]: Ah, of course because I am using the latest spikeinterface version. If you don't have it, your array will be loaded as an object that will have Nan.

**Comment by pauladkisson** [comment]
> > > In short: I want to have the bug solved and I want to think about the complicated case more. They are two different issues so I would prefer them to address them piece-wise.
> > 
> > 
> > Ok, I am conviced. 👍
> 
> Thanks but man I feel bad about my lack of clear thinking on this issue. I have made too many requests as I have changed my mind and learned thinking about this. I can take care of rebasing and organizing this PR so you don't have to if you want.

No worries.

**Review by h-mayorquin — APPROVED** [review]
Thanks again for all the work.
